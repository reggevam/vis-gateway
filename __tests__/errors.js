const { createTestClient } = require('apollo-server-testing');
const { gql } = require('apollo-server');
const { constructTestServer } = require('../__testUtils');
require('dotenv').config();

const unknownErrorMutation = gql`
  mutation throwError {
    throwError
  }
`;

const fileQuery = gql`
  query file($id: ID!) {
    file(id: $id) {
      filename
      mimetype
      encoding
      id
      content
    }
  }
`;

const errorReportMutation = gql`
  mutation errorReport($message: String!) {
    errorReport(message: $message)
  }
`;

describe('errors', () => {
  beforeEach(() => {
    console.error = jest.fn();
  });

  afterEach(() => {
    console.error.mockClear();
  });
  afterAll(() => {
    console.error.mockRestore();
  });

  test.each`
    title               | fragment                | type          | variables
    ${'known error'}    | ${fileQuery}            | ${'query'}    | ${{ id: 'nope...' }}
    ${'un-known error'} | ${unknownErrorMutation} | ${'mutation'} | ${{ id: 'nope...' }}
    ${'client error'}   | ${errorReportMutation}  | ${'mutation'} | ${{ message: 'something has gone wrong in the client' }}
  `('$title', async ({ fragment, type, variables }) => {
    const { server } = constructTestServer();
    const testClient = createTestClient(server);
    if (type === 'query') {
      const { query: clientQuery } = testClient;
      await clientQuery({ query: fragment, variables });
    } else {
      const { mutate } = testClient;
      await mutate({ mutation: fragment, variables });
    }
    expect(console.error).toHaveBeenCalledTimes(1);
    expect(console.error.mock.calls[0][0]).toMatchSnapshot();
  });
});
