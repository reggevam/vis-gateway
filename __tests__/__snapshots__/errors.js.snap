// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`errors client error 1`] = `
Object {
  "extensions": Object {
    "code": "CLIENT_ERROR",
  },
  "locations": Array [
    Object {
      "column": 3,
      "line": 2,
    },
  ],
  "message": "something has gone wrong in the client",
  "path": Array [
    "errorReport",
  ],
}
`;

exports[`errors known error 1`] = `
Object {
  "extensions": Object {
    "code": "FILE_NOT_FOUND",
  },
  "locations": Array [
    Object {
      "column": 3,
      "line": 2,
    },
  ],
  "message": "could not find fileId nope...",
  "path": Array [
    "file",
  ],
}
`;

exports[`errors un-known error 1`] = `
Object {
  "extensions": Object {
    "code": "GRAPHQL_VALIDATION_FAILED",
  },
  "locations": Array [
    Object {
      "column": 3,
      "line": 2,
    },
  ],
  "message": "Cannot query field \\"throwError\\" on type \\"Mutation\\".",
  "path": undefined,
}
`;
